include/drivers/sensor.h

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright © 2017-2018 The Crust Firmware Authors.
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#ifndef DRIVERS_SENSOR_H
+#define DRIVERS_SENSOR_H
+
+#include <dm.h>
+#include <intrusive.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <util.h>
+
+#define SENSOR_OPS(dev) \
+	(&container_of((dev)->drv, struct sensor_driver, drv)->ops)
+
+struct sensor_conversion {
+	const int16_t min_value;  /**< Addition factor. */
+	const int16_t multiplier; /**< Multiplicaton factor. */
+};
+
+struct sensor_info {
+	const char *const              name;    /**< Name exported to SCPI. */
+	uint32_t                       reading; /**< Current value from sensor.
+	                                         * */
+	const struct sensor_conversion formula; /**< Information on converting
+	                                         * cooked value. */
+	const uint8_t                  address; /**< address of register. */
+};
+
+struct sensor_driver_ops {
+	struct sensor_info *(*get_info)(struct device *dev, uint8_t id);
+	uint32_t            (*read_raw)(struct device *dev, uint8_t id,
+	                                uint32_t *raw);
+};
+
+struct sensor_driver {
+	const struct driver            drv;
+	const struct sensor_driver_ops ops;
+};
+
+/**
+ * Get the number of sensors within a device.
+ *
+ * This function has no defined errors.
+ *
+ * @param dev   A device containing one or more sensors.
+ * @return      The number of sensors available on this device.
+ */
+
+/**
+ * Get generic information about a sensor.
+ *
+ * This function has no defined errors.
+ *
+ * @param dev   The device containing this sensor.
+ * @param id    The device-specific identifier for this sensor.
+ * @return      A pointer to the information structure.
+ */
+static inline struct sensor_info *
+sensor_get_info(struct device *dev, uint8_t id)
+{
+	return SENSOR_OPS(dev)->get_info(dev, id);
+}
+
+/**
+ * Get the current reading from a sensor. If the value is unknown, this
+ * function returns UINT32_MAX for the value
```

Comment:
This function should use the same error codes as everywhere else

Reply:
Comment was written before the function. The function implementation does use the same error codes as everywhere else. This comment is now fixed.